### ISSUE TITLE
completecases for dataframe without allowing for missing could be no-op

### ISSUE THREAD

**Issue by Ken-B** [JuliaData/DataFrames.jl#1358]
tl;dr Couldn't `complete_cases` just check for each column whether it allows missing values and only perform the loop (that checks for each element if it's missing) if this is the case? 
That would mean that for a dataframe without columns that allow for missing values, `completecases` could just return all `trues` and `dropmissing` would be an no-op.

If you agree I could look into making a PR. I would needs some help as I think it would require using type trickery, which I'm not so experienced with.

_Background_: I found out that a significant part of my calculation (using `FixedEffectModels.jl`) was spent in `completecases`.
Initially I found that categorical columns use much more memory and are about 10x slower than non-categorical. Here's a MWE in julia v0.6.
```
using DataFrames
data = rand(1:100, 1_000_000);
df = DataFrame(data=data);
df[:datacat] = categorical(df[:data]);

completecases(df[[:data]]); #compile
@time cc = completecases(df[[:data]]);

completecases(df[[:datacat]]); #compile
@time cc = completecases(df[[:datacat]]);

#OUTPUT
#0.002060 seconds (44 allocations: 124.656 KiB)
#0.030504 seconds (1.00 M allocations: 30.639 MiB, 66.29% gc time)
```
For columns that allow for missing data, it's even 10x worse (categorical with missing is fine again - weird).
```
df[:mdata] = allowmissing(data);
completecases(df[[:mdata]]);
@time cc = completecases(df[[:mdata]]);

df[:mdatacat] = categorical(df[:mdata]);
completecases(df[[:mdatacat]]);
@time cc = completecases(df[[:mdatacat]]);
#0.376526 seconds (3.00 M allocations: 61.141 MiB, 6.63% gc time)
#0.002844 seconds (48 allocations: 124.781 KiB)
```
However, in julia v0.7 all these differences go away and none of them allocate. Tip of the hat for Julia 1.0! :)

**Comment by nalimilan** [comment]
The best place to handle this would be the compiler, which should be able to optimize `any(ismissing, x)` calls when `ismissing` is always false. Currently that doesn't work because of bounds checks, see https://github.com/JuliaLang/julia/issues/21256.

I guess in the meantime it would be OK to check `eltype(x) >: Missing` before calling `any(ismissing, x)`, with a comment pointing to this issue. AFAICT, `dropmissing` cannot be a no-op, though, since it needs to return a copy in all cases.

**Comment by Ken-B** [comment]
Ok, 2 things. 
First, it seems that the slowdown in Julia v0.6 comes not only from the `ismissing`, but also from the `enumerate`. If in `_nonmissing!` I replace `for (i, el) in enumerate(col)` with `for i in eachindex(res, col)`, the extra allocations go away (and speedup x50). 
However, this seems to be fixed in Julia v0.7, so is there a separate branch that tracks Julia v0.6? Does it make sense to make this change and then change it back for v0.7?

Secondly, following your elegant suggestion with `eltype(x) >: Missing`, only a simple change would be needed to fix this issue:
```julia
function _nonmissing!(res, col)
    eltype(col) >: Missing || return
    @inbounds for i in eachindex(res, col)
        res[i] &= !ismissing(col[i])
    end
end
```
Feel free to push the change yourself, but I can also easily make a PR, as you wish.

**Comment by nalimilan** [comment]
At this point I wouldn't make changes which make things faster only on 0.6. But please make a PR for the other change.
